design-docs/opmodes-commandbased.md

```diff
@@ -0,0 +1,205 @@
+  }
+
+  // exposed so users may override it if desired, but generally shouldn't be necessary to do so
+  public void robotPeriodic() {
+    beforeScheduler();
+    CommandScheduler.run();
+    afterScheduler();
+  }
+}
+```
+
+## Subsystem enhancements
+
+The `Subsystem` interface adds an overload of `setDefaultCommand` to support creating per-opmode default commands, and similarly a new `removeDefaultCommand` overload.  The non-`CommandOpMode` overloads are applied in all opmodes where a per-opmode default has not been set.
+
+```java
+class Subsystem {
+  // Sets a default command for this subsystem active only in the given opmode.
+  default void setDefaultCommand(CommandOpMode opmode, Command defaultCommand);
+
+  // Sets a default command for this subsystem active in all opmodes where there's no per-opmode default command.
+  default void setDefaultCommand(Command defaultCommand);
+
+  // Removes the default command for the given opmode.  No effect if none set.
+  default void removeDefaultCommand(CommandOpMode opmode);
+
+  // Removes the default command for all opmodes where there's no per-opmode default command.
+  default void removeDefaultCommand();
+}
+```
+
+## Java Robot Code Examples
+
+The template/example code for command-based Java includes the following:
+- A Robot class with subsystems and constructor that sets up a command-based teleop opmode, a couple of auto opmodes, and a test opmode
+
+Robot:
+
+```java
+public class Robot extends CommandRobot {
+  // actuators (Subsystem derived)
+  public final Drive drive = new Drive();
+  public final Intake intake = new Intake();
+  public final Storage storage = new Storage();
+
+  // sensors (not Subsystem derived)
+  public final Vision vision = new Vision();
+
+  public Robot() {
+    // Automatically disable and retract the intake whenever the ball storage is full.
+    storage.hasCargo.onTrue(intake.retractCommand());
+
+    // Create auto opmodes
+    addSimpleAuto();
+    addPathAuto("drive and turn");
+
+    // Create teleop opmodes
+    addArcadeTeleop();
+  }
+
+  private void addSimpleAuto() {
+    // A simple autonomous opmode
+    CommandOpModes.autonomous("Simple Auto").running.whileTrue(Autos.simpleAuto(this));
+  }
+
+  private void addPathAuto(String path) {
+    // A complex autonomous opmode that loads a path when selected in the DS while still disabled
+    CommandOpMode opmode = CommandOpModes.autonomous(path, "Follow Path");
+    opmode.selected.onTrue(Commands.runOnce(() -> Paths.loadPath(path)));
+    opmode.running.whileTrue(Autos.followPath(this, path));
+  }
+
+  private void addArcadeTeleop() {
+    // A teleop opmode with joystick and button controls
+    CommandOpMode opmode = CommandOpModes.teleoperated("teleop");
+
+    var driverController = new CommandXboxController(1);
+
+    // Control the drive with split-stick arcade controls
+    m_drive.setDefaultCommand(
+        opmode,
```

Comment:
This is extremely verbose and often pointless to specify since operator controls are in teleop opmodes. Can’t there be a way to add commands per-opmode if defined like
`opmode.add(controller.x().whileTrue(()->drive.drive(…)),…)` and default to all of the teleop/test opmodes. This way seems less verbose and more structured.

Reply:
How would the implementation of that work?  It might be possible by some way injecting an additional condition into the trigger creation/binding behind the scenes, but that feels somewhat fragile.

Reply:
It could work if the Triggers were bound to an opmode inside of the class which can be checked in the command scheduler periodic?

The opmode can call a `Trigger.setOpMode(OpMode)` inside the `OpMode.add` function. Triggers can be constructed with an optional opmode filter which stores the opmode(s) it runs under.